### ISSUE TITLE
Issues downloading and installing new OS versions for raspberry-pi

### ISSUE THREAD

**Issue by mscharley** [resin-io/resin-cli#452]
I'm having issues with provisioning new devices:

```
$ resin device init -va Foo
Getting device operating system for raspberry-pi
Downloading Device OS [========================] 100% eta 39s  
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: incorrect data check
    at Zlib._handle.onerror (zlib.js:370:17)
```

> resin-cli: 5.5.0
> node: 6.10.0
> npm: 3.10.10

**Comment by mscharley** [comment]
The images that are downloaded are valid .img files - I can mount them locally completely fine.

**Comment by lekkas** [comment]
Hey @mscharley , 

You can also track https://github.com/resin-io/resin-cli/issues/451 , which was opened as a response to your issue. Cheers!

**Comment by emirotin** [comment]
@lekkas so I'm currently working on making the OS v1 downloadable via the CLI as requested in #451, but it seems that the rest of the CLI would still not work with the OS v2 downloads (as highlighted by this issue). Am I right?

**Comment by emirotin** [comment]
Hey @mscharley I've just published the CLI v5.6.0, can you give it a try?

Here's the full story behind the issue and what was done:
- we've recently published two preview releases of the resinOS v2.0.0 for the RPi family
- the CLI was always downloading the most recent release which always used to be v1 until recently
- it looks (I didn't have time to investigate it yet) the CLI configuration doesn't work with OS v2 yet
- so you got the new OS image and configuration failed
- I've made 2 changes: first, you can now specify the version, check `resin help os download` for the variety of options, second, the default version is the most recent **non-pre-release** now, which would be the most recent v1 for RPi which is what it was previously

So now if you try to run the `device init` it will download the OS v1 and _should_ work.
Please let us know if it's the case.

But we still need to check if the OS configuration actually works with resinOS v2, and fix it if not.

**Comment by mscharley** [comment]
`resin device init` is now working fine with 5.6.0 to install OS v1.24.1

A minor point of feedback though that's only slightly tangential - it'd be great if the CLI had an option to select which version as one of the options in the wizard if no option was given explicitly.

Currently our application isn't compatible with the 2.x OS either, leading to redirect loops if it ever got flashed.

**Comment by emirotin** [comment]
Great to know!

Yeah, the OS version selection can and will be added to the wizard, so far I've implemented all the required primitives.

**Comment by mscharley** [comment]
Actually, I spoke too soon.

Now I'm having a different issue entirely:

```
$ resin login
______          _         _
| ___ \        (_)       (_)
| |_/ /___  ___ _ _ __    _  ___
|    // _ \/ __| | '_ \  | |/ _ \
| |\ \  __/\__ \ | | | |_| | (_) |
\_| \_\___||___/_|_| |_(_)_|\___/

Logging in to resin.io
? How would you like to login? Web authorization (recommended)
Connecting to the web dashboard
Successfully logged in as: sysadmin

Now what?

Run the following command to get a device started with Resin.io

  $ resin quickstart

Find out about more super powers by running:

  $ resin help

If you need help, or just want to say hi, don't hesitate in reaching out at:

  GitHub: https://github.com/resin-io/resin-cli/issues/new
  Gitter: https://gitter.im/resin-io/chat
$ resin device init -va DocketPrinterStage
Getting device operating system for raspberry-pi
OS version is not specified, using the default version: the newest stable (non-pre-release) version if available, or the newe
st version otherwise (if all existing versions for the given device type are pre-release).
The image was downloaded successfully
Registering to DocketPrinterStage: dd2684ce208b1357a4bee73eec399b23b9fad29a3aac02b90c1fd60d49172a
Configuring operating system image
? Network Connection ethernet
? Check for updates every X minutes 5
Initializing a device requires administrative permissions
given that we need to access raw devices directly.

Type your computer password to continue
Password:
You have to log in to continue

Run the following command to go through the login wizard:

  $ resin login

If you need help, don't hesitate in contacting us at:

  GitHub: https://github.com/resin-io/resin-cli/issues/new
  Gitter: https://gitter.im/resin-io/chat

Done
```

**Comment by emirotin** [comment]
But what is exactly the problem? Auth was lost between the attempts?

**Comment by emirotin** [comment]
Hm, actually I see it little bit differently. When it asks me to type the computer password _I think_ it should show the system dialog, but instead nothing happens and as I type the character in the CLI it advances to the next question.

**Comment by emirotin** [comment]
UPD: so no, it's not supposed to show the system dialog, it's supposed to run the sudo command but something goes wrong, after a short timeout the control returns. 

I think one contribution here is that I have passwordless sudo for this account to simplify some testing and that might be the issue. Will test it separately and file a new issue.

**Comment by emirotin** [comment]
So yes, I've created a separate issue for that and now I can see your problem: err'ing about the user not being logged in.

**Comment by lekkas** [comment]
> but it seems that the rest of the CLI would still not work with the OS v2 downloads (as highlighted by this issue. Am I right?

Yes you are right @emirotin , it's possible that there are other subcommands that depend no OS v1 specifics . 

> the default version is non-pre-release now, which would be the most recent v1 for RPi which is what it was previously

+1 on this. How will we handle v2 'proper' (non-prerelease) versions though? Hm I remember a thread a few days ago were we discussed this internally were the `strings` method was mentioned right?

**Comment by emirotin** [comment]
`strings` is a dev-time hack for quickly checking the downloaded version, probably doesn't work well for production
We need a way to detect the version reliably (Petros has mentioned the presence of some file, I will find it), and we need to understand what's the problem with v2 and what should be changed. I'll create a separate issue for this, let's keep this for the v1 happy path that's currently broken.

**Comment by emirotin** [comment]
So back to the original issue, I've confirmed my guess but I wonder how that used to work before.
The `device init` command sequences a bunch of other: it first downloads the image, then configures it (injects the app- and device-specific config), and then writes to the SD card.

The config injection involves some commands like mounting the img file that have to be ran as `root`. When later the CLI tries to check if the user is logged in it looks for the `~/.resin/token` file. But as the command is run by `sudo` it the home folder now is (normally) `/root` and not the user's home folder, so there's no saved token in there.

I've tested it back to `v4.5.0` and can see the very same error.
@mscharley did it work for you before? Can you somehow check the version that used to work (it won't now because of the OS v2, but I'm just curious which version it was if any)?

Now for the wrokaround, @mscharley I believe if you do (assuming Linux or macOS)
```
sudo mkdir -p /root/.resin && sudo cp ~/.resin/token /root/.resin/
```

This is a hacky way and we're working on the proper fix, but that should get you unblocked right now.

**Comment by mscharley** [comment]
Why do you need to do `sudo resin <something>` at all? Feels like that's asking for trouble for all sorts of reasons, the least of which is auth tokens. Running eg. `sudo dd whatever` or `sudo mount` directly would be much more reliable.

**Comment by emirotin** [comment]
I don't quite get the part about `Running eg. sudo dd whatever or sudo mount directly would be much more reliable.` Do you suggest interrupting the flow and asking the user to mount something before continuing?

I didn't work on the CLI before recently, but the core idea of the quickstart / provision flow was to have the most frictionless process for the end user. So the CLI tries to automate and abstract away the things as much as possible.

I don't know yet which exact part of the flow needs the admin permissions, but it's either mounting / writing to the disk image / unmounting.

If you're cautious about the security of these operations you can always use the lower-level methods (`os download` / `os initialize` <- this needs sudo / and then manual unmount).
The `device init` shortcut was designed to streamline the process, but that's why it requires the `sudo` permissions at some point.

Thanks for the version number, it's really an old one, I will check how it used to solve the init problem without trouble (do you say you didn't have to provide the sudo password there?)

**Comment by mscharley** [comment]
I don't mean that you should ask the user to run stuff, run it as part of the resin CLI process, but instead of trying to run `sudo resin [whatever subcommand]` which will run Resin as root and introduce the issues, run the commands that `resin [whatever subcommand]` would run with `sudo` in front of them so, eg.

Instead of:

```
- resin init
  - sudo resin subcommand
    - dd foo bar
    - some other subcommand
```

Do the following:

```
- resin init
  - resin subcommand
    - sudo dd foo bar
    - sudo some other subcommand
```

**Comment by emirotin** [comment]
Oh, now I get you @mscharley. Yes, this is my actual plan for the fix to this problem.

**Comment by emirotin** [comment]
Hey @mscharley sorry it really took longer than I've expected because after fixing the sudo issue I've found another one that involved fixing and updating 2 sub-dependencies.

But now v5.7.0 should work properly (at least it does in my tests). Please let me know if it works or reopen the issue otherwise.
